vsphere/resource_vsphere_vmfs_datastore.go

```diff
@@ -0,0 +1,324 @@
+		Update: resourceVSphereVmfsDatastoreUpdate,
+		Delete: resourceVSphereVmfsDatastoreDelete,
+		Schema: s,
+	}
+}
+
+func resourceVSphereVmfsDatastoreCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*govmomi.Client)
+	hsID := d.Get("host_system_id").(string)
+	dss, err := hostDatastoreSystemFromHostSystemID(client, hsID)
+	if err != nil {
+		return fmt.Errorf("error loading host datastore system: %s", err)
+	}
+
+	// To ensure the datastore is fully created with all the disks that we want
+	// to add to it, first we add the initial disk, then we expand the disk with
+	// the rest of the extents.
+	disks := d.Get("disks").([]interface{})
+	disk := disks[0].(string)
+	spec, err := diskSpecForCreate(dss, disk)
+	if err != nil {
+		return err
+	}
+	spec.Vmfs.VolumeName = d.Get("name").(string)
+	ctx, cancel := context.WithTimeout(context.Background(), defaultAPITimeout)
+	defer cancel()
+	ds, err := dss.CreateVmfsDatastore(ctx, *spec)
+	if err != nil {
+		return fmt.Errorf("error creating datastore with disk %s: %s", disk, err)
+	}
+
+	// Now add any remaining disks.
+	for _, disk := range disks[1:] {
+		spec, err := diskSpecForExtend(dss, ds, disk.(string))
+		if err != nil {
+			// We have to destroy the created datastore here.
+			if remErr := removeDatastore(dss, ds); remErr != nil {
+				// We could not destroy the created datastore and there is now a dangling
+				// resource. We need to instruct the user to remove the datastore
+				// manually.
+				return fmt.Errorf(formatCreateRollbackErrorUpdate, disk, err, remErr)
+			}
+			return fmt.Errorf("error fetching datastore extend spec for disk %s: %s", disk, err)
+		}
+		ctx, cancel := context.WithTimeout(context.Background(), defaultAPITimeout)
+		defer cancel()
+		if _, err := extendVmfsDatastore(ctx, dss, ds, *spec); err != nil {
+			if remErr := removeDatastore(dss, ds); remErr != nil {
+				// We could not destroy the created datastore and there is now a dangling
+				// resource. We need to instruct the user to remove the datastore
+				// manually.
+				return fmt.Errorf(formatCreateRollbackErrorUpdate, disk, err, remErr)
+			}
+			return fmt.Errorf("error extending datastore with disk %s: %s", disk, err)
+		}
+	}
+
+	d.SetId(ds.Reference().Value)
+
+	// Done
+	return resourceVSphereVmfsDatastoreRead(d, meta)
+}
+
+func resourceVSphereVmfsDatastoreRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*govmomi.Client)
+	id := d.Id()
+	ds, err := datastoreFromID(client, id)
+	if err != nil {
+		return fmt.Errorf("cannot find datastore: %s", err)
+	}
+	props, err := datastoreProperties(ds)
+	if err != nil {
+		return fmt.Errorf("could not get properties for datastore: %s", err)
+	}
+	if err := flattenDatastoreSummary(d, &props.Summary); err != nil {
+		return err
+	}
+
+	// We also need to update the disk list from the summary.
+	var disks []string
+	for _, disk := range props.Info.(*types.VmfsDatastoreInfo).Vmfs.Extent {
```

Comment:
there's a ton of potential crashes we're not tackling here? I've seen use use flatten methods to track this in other providers

Reply:
Since we are dealing in VMFS datastores only here, this is guaranteed to be `types.VmfsDatastoreInfo`. So if the assertion fails, this is a breaking bug that needs to be looked at ASAP, so I'm thinking the crash is warranted here.